lib/dialects/abstract/connection-manager.js

```diff
@@ -265,31 +265,23 @@ class ConnectionManager {
       promise = Promise.resolve();
     }
 
-    return promise.then(() =>
-      new Promise((resolve, reject) => {
-        const connectionPromise = this.pool.acquire(options.priority, options.type, options.useMaster);
-        const connectionTimer = timers.setInterval(() => {
-          let evictTimer = false;
-
-          if (connectionPromise.isFulfilled()) {
-            resolve(connectionPromise);
-            debug('connection acquire');
-            evictTimer = true;
-          } else if (this.poolError) {
+    return promise.then(() => {
+      return Promise.race([
+        this.pool.acquire(options.priority, options.type, options.useMaster),
+        new Promise((resolve, reject) =>
+        timers.setTimeout(() => {
+          if (this.poolError) {
             reject(this.poolError);
-            this.poolError = null;
-            evictTimer = true;
-          } else if (connectionPromise.isRejected()) {
-            connectionPromise.catch(reject);
-            evictTimer = true;
           }
-
-          if (evictTimer) {
-            timers.clearInterval(connectionTimer);
-          }
-        }, 0);
-      })
-    );
+        }, 0))
```

Comment:
> What guarantees you that any possible acquire error would have been set in the next event loop tick?

If there is an error in pooling lib, then it would be emitted. But yeah there is no guarantee that order of error will be preserved. Because its all up to generic-pool to emit this error, which is not emitted per acquire call level, but rather per pool level.

To properly solve this issue we have to track all acquire calls in a queue and check which one failed to acquire connection and reject that. In other words we will need our own pooling, hence removing generic-pool from play

`this.poolErr` is not safe nor does `pool.on('factoryCreateError')`